clone.go

```diff
@@ -0,0 +1,192 @@
+package poly
+
+import (
+	"errors"
+	"regexp"
+	"sort"
+	"strings"
+	"sync"
+)
+
+type CloneSequence struct {
+	Sequence string
+	Circular bool
+}
+
+type Overhang struct {
+	Length   int
+	Position int
+	Forward  bool
+}
+
+type Fragment struct {
+	Sequence        string
+	ForwardOverhang string
+	ReverseOverhang string
+}
+
+type Enzyme struct {
+	EnzymeName        string
+	RegexpFor         *regexp.Regexp
+	RegexpRev         *regexp.Regexp
+	EnzymeSkip        int
+	EnzymeOverhangLen int
+	Directional       bool
+}
+
+/******************************************************************************
+
+Base cloning functions begin here.
+
+******************************************************************************/
+
+// https://qvault.io/2019/10/21/golang-constant-maps-slices/
+func getBaseRestrictionEnzymes() map[string]Enzyme {
+	// Eventually, we want to get the data for this map from ftp://ftp.neb.com/pub/rebase
+	enzymeMap := make(map[string]Enzyme)
+
+	// Build default enzymes
+	enzymeMap["BsaI"] = Enzyme{"BsaI", regexp.MustCompile("GGTCTC"), regexp.MustCompile("GAGACC"), 1, 4, true}
+	enzymeMap["BbsI"] = Enzyme{"BbsI", regexp.MustCompile("GAAGAC"), regexp.MustCompile("GTCTTC"), 2, 4, true}
+	enzymeMap["BtgZI"] = Enzyme{"BtgZI", regexp.MustCompile("GCGATG"), regexp.MustCompile("CATCGC"), 10, 4, true}
+	enzymeMap["BsmBI"] = Enzyme{"BsmBI", regexp.MustCompile("CGTCTC"), regexp.MustCompile("GAGACG"), 1, 4, true}
+
+	// Return EnzymeMap
+	return enzymeMap
+}
+
+func RestrictionEnzymeCut(seq CloneSequence, enzymeStr string) ([]Fragment, error) {
+	enzymeMap := getBaseRestrictionEnzymes()
+	if _, ok := enzymeMap[enzymeStr]; ok == false {
+		return []Fragment{}, errors.New("Enzyme " + enzymeStr + " not found in enzymeMap")
+	}
+	enzyme, _ := enzymeMap[enzymeStr]
+	return RestrictionEnzymeCutEnzymeStruct(seq, enzyme), nil
+}
+
+func RestrictionEnzymeCutEnzymeStruct(seq CloneSequence, enzyme Enzyme) []Fragment {
+	var fragmentSeqs []string
+	var sequence string
+	if seq.Circular == true {
+		sequence = strings.ToUpper(seq.Sequence + seq.Sequence)
+	} else {
+		sequence = strings.ToUpper(seq.Sequence)
+	}
+	// Find and define overhangs
+	var overhangs []Overhang
+	forwardCuts := enzyme.RegexpFor.FindAllStringIndex(sequence, -1)
+	reverseCuts := enzyme.RegexpRev.FindAllStringIndex(sequence, -1)
+	for _, forwardCut := range forwardCuts {
+		overhangs = append(overhangs, Overhang{Length: enzyme.EnzymeOverhangLen, Position: forwardCut[1] + enzyme.EnzymeSkip, Forward: true})
+	}
+	for _, reverseCut := range reverseCuts {
+		overhangs = append(overhangs, Overhang{Length: enzyme.EnzymeOverhangLen, Position: reverseCut[0] - enzyme.EnzymeSkip, Forward: false})
+	}
+	// TODO add checks that make sure overhangs + enzymeSkip is over or below the length of the string
```

Comment:
Do these checks needs to be added before merge? Will this cause silent fails and bad plasmid builds?